pkg/plugins/ifaces.go

```diff
@@ -125,3 +126,7 @@ type ClientMiddlewareFunc func(next Client) Client
 func (fn ClientMiddlewareFunc) CreateClientMiddleware(next Client) Client {
 	return fn(next)
 }
+
+type OAuth2Service interface {
+	SaveExternalService(ctx context.Context, cmd *oauthserver.ExternalServiceRegistration) (*oauthserver.ClientDTO, error)
```

Comment:
Ideally `pkg/plugins` would have their own copy of the structs coming from `oauthserver` if possible 🙏

Reply:
is it okay to duplicate code or should the service import the structs from here? Also, in the end, `wire.go` links the minimal interface with the full structure, so is there any benefit of doing this apart than having an "isolated" `pkg/plugin`?

Reply:
Duplicate is fine yeah 😄 Yeah re Wire that's also fine, we do that with a few plugins things. It's really mostly about the isolated package more than anything for now

Reply:
I had to remove the duplicated code from `pkg/services/oauthserver` since if not, the implementation didn't match the interface (since both were pointing to different structs)

Reply:
yes :+1: that's what I did (If I am understanding you correctly)

Reply:
Oh, but now we have a new friend `"github.com/grafana/grafana/pkg/services/accesscontrol"` being imported 😉

Reply:
this is getting quite messy, are you sure this is the right direction? I don't think we should redefine (or move) _only_ the `Permission` struct here and change it everywhere where all the context about these structs are in `pkg/services`